File: ChatUI/src/core/simpleEvents.js
/**
 * Simple Events System
 * Handles our core event types without unnecessary complexity
 */

// Simple event types
export const SIMPLE_EVENT_TYPES = {
  CHAT_MESSAGE: "chat_message",
  ROUTE_TO_ARTIFACT: "route_to_artifact",
  ROUTE_TO_CHAT: "route_to_chat", 
  UI_TOOL_ACTION: "ui_tool_action",
  STATUS: "status",
  ERROR: "error"
};

/**
 * Simple Event Processor
 * Processes our simplified events
 */
export class SimpleEventProcessor {
  constructor(artifactManager, chatInterface, config = {}) {
    this.artifactManager = artifactManager;
    this.chatInterface = chatInterface;
    this.config = config;
    
    console.log('✅ Simple Event Processor initialized');
  }

  /**
   * Process incoming simple event
   */
  processEvent(event) {
    try {
      const parsedEvent = typeof event === 'string' ? JSON.parse(event) : event;
      
      // Validate event has required fields
      if (!parsedEvent.type) {
        console.warn('Invalid event: missing type', parsedEvent);
        return;
      }
      
      switch (parsedEvent.type) {
        case SIMPLE_EVENT_TYPES.CHAT_MESSAGE:
          this.handleChatMessage(parsedEvent);
          break;
          
        case SIMPLE_EVENT_TYPES.ROUTE_TO_ARTIFACT:
          this.handleArtifactRoute(parsedEvent);
          break;
          
        case SIMPLE_EVENT_TYPES.ROUTE_TO_CHAT:
          this.handleInlineComponentRoute(parsedEvent);
          break;
          
        case SIMPLE_EVENT_TYPES.UI_TOOL_ACTION:
          this.handleUIToolAction(parsedEvent);
          break;
          
        case SIMPLE_EVENT_TYPES.STATUS:
          this.handleStatus(parsedEvent);
          break;
          
        case SIMPLE_EVENT_TYPES.ERROR:
          this.handleError(parsedEvent);
          break;
          
        default:
          console.log('Unknown event type:', parsedEvent.type);
          // Pass through to chat interface
          this.chatInterface.addMessage?.({
            sender: 'system',
            content: `Unknown event: ${parsedEvent.type}`,
            timestamp: Date.now()
          });
          break;
      }
    } catch (error) {
      console.error('Error processing simple event:', error, event);
    }
  }

  /**
   * Handle chat message
   */
  handleChatMessage(event) {
    const { content, sender, role } = event.data;
    
    this.chatInterface.addMessage?.({
      content: content,
      sender: sender || 'assistant',
      role: role || 'assistant',
      timestamp: event.timestamp || Date.now()
    });
  }

  /**
   * Handle artifact component routing (full-featured right panel components)
   */
  handleArtifactRoute(event) {
    const { title, content, category, artifact_id, component_name, component_data } = event.data;
    
    // Create artifact in artifact manager with component info
    this.artifactManager.createArtifact?.({
      id: artifact_id,
      title: title,
      content: content,
      category: category || 'general',
      componentName: component_name,
      componentData: component_data,
      needsWorkflowResolution: true,
      timestamp: event.timestamp || Date.now()
    });
    
    console.log(`📄 Created artifact component: ${component_name || title}`);
  }

  /**
   * Handle inline component routing (lightweight UI elements in chat)
   */
  handleInlineComponentRoute(event) {
    const { content, component_name, component_data } = event.data;
    
    // Add as chat message with inline component
    this.chatInterface.addMessage?.({
      content: content,
      sender: 'assistant',
      type: 'inline_component',
      inlineComponent: {
        name: component_name,
        data: component_data,
        // Let the UI loader resolve the actual component from workflow
        needsWorkflowResolution: true
      },
      timestamp: event.timestamp || Date.now()
    });
    
    console.log(`🔧 Routed to inline component: ${component_name}`);
  }

  /**
   * Handle UI tool action
   */
  handleUIToolAction(event) {
    const { tool_id, action_type, payload } = event.data;
    
    // Emit UI tool action for components to handle
    if (this.config.onUIToolAction) {
      this.config.onUIToolAction(tool_id, action_type, payload);
    }
    
    console.log(`🔧 UI Tool Action: ${tool_id} - ${action_type}`);
  }

  /**
   * Handle status message
   */
  handleStatus(event) {
    const { message } = event.data;
    
    this.chatInterface.addMessage?.({
      content: message,
      sender: 'system',
      type: 'status',
      timestamp: event.timestamp || Date.now()
    });
  }

  /**
   * Handle error message
   */
  handleError(event) {
    const { message, code } = event.data;
    
    this.chatInterface.addMessage?.({
      content: `Error: ${message}`,
      sender: 'system',
      type: 'error',
      timestamp: event.timestamp || Date.now()
    });
    
    console.error('Event error:', message, code);
  }
}

/**
 * React Hook for Simple Event Processing
 */
export function useSimpleEvents(transport, artifactManager, chatInterface, config = {}) {
  const [processor, setProcessor] = React.useState(null);

  React.useEffect(() => {
    if (!transport || !artifactManager || !chatInterface) return;

    const eventProcessor = new SimpleEventProcessor(artifactManager, chatInterface, config);
    setProcessor(eventProcessor);

    // Set up transport event listeners
    const handleTransportMessage = (data) => {
      eventProcessor.processEvent(data);
    };

    // Listen for events from transport
    transport.addEventListener?.('message', handleTransportMessage);
    transport.onmessage = handleTransportMessage; // Fallback for different transport types

    return () => {
      transport.removeEventListener?.('message', handleTransportMessage);
      transport.onmessage = null;
    };
  }, [transport, artifactManager, chatInterface]);

  return {
    processor
  };
}

/**
 * Simple Action Handler for sending actions back to backend
 */
export class SimpleActionHandler {
  constructor(transport) {
    this.transport = transport;
  }

  async sendChatMessage(content) {
    await this.transport.send({
      type: SIMPLE_EVENT_TYPES.CHAT_MESSAGE,
      data: { content, sender: 'user', role: 'user' }
    });
  }

  async sendUIToolAction(toolId, actionType, payload) {
    await this.transport.send({
      type: SIMPLE_EVENT_TYPES.UI_TOOL_ACTION,
      data: { tool_id: toolId, action_type: actionType, payload }
    });
  }
}

export default {
  SimpleEventProcessor,
  useSimpleEvents,
  SimpleActionHandler,
  SIMPLE_EVENT_TYPES
};

File: core/events/simple_events.py
# ==============================================================================
# FILE: core/ui/simple_events.py
# DESCRIPTION: Simplified event system
# ==============================================================================

from typing import Dict, Any, Optional
from dataclasses import dataclass
from enum import Enum
import json
import time
import logging

from logs.logging_config import get_event_logger

event_logger = get_event_logger("simple_events")


class SimpleEventType(str, Enum):
    """Simple event types for our actual needs"""
    CHAT_MESSAGE = "chat_message"
    ROUTE_TO_ARTIFACT = "route_to_artifact" 
    ROUTE_TO_CHAT = "route_to_chat"
    UI_TOOL_ACTION = "ui_tool_action"
    STATUS = "status"
    ERROR = "error"


@dataclass
class SimpleEvent:
    """Base event class"""
    type: SimpleEventType
    data: Dict[str, Any]
    timestamp: Optional[int] = None
    agent_name: Optional[str] = None
    
    def __post_init__(self):
        if self.timestamp is None:
            self.timestamp = int(time.time() * 1000)
    
    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary for JSON serialization"""
        result = {
            "type": self.type,
            "data": self.data,
            "timestamp": self.timestamp
        }
        if self.agent_name:
            result["agent_name"] = self.agent_name
        return result
    
    def to_json(self) -> str:
        """Convert to JSON string"""
        return json.dumps(self.to_dict())


class SimpleEventEncoder:
    """Simple event encoder"""
    
    def encode_sse(self, event: SimpleEvent) -> str:
        """Encode event for SSE transport"""
        return f"data: {event.to_json()}\n\n"
    
    def encode_websocket(self, event: SimpleEvent) -> str:
        """Encode event for WebSocket transport"""
        return event.to_json()


# Event creation helpers
def create_chat_message(content: str, sender: str, role: str = "assistant") -> SimpleEvent:
    """Create a simple chat message event"""
    event_logger.debug("Creating chat message event", extra={
        "sender": sender,
        "role": role,
        "content_length": len(content) if content else 0
    })
    
    return SimpleEvent(
        type=SimpleEventType.CHAT_MESSAGE,
        data={
            "content": content,
            "sender": sender,
            "role": role
        },
        agent_name=sender
    )


def create_artifact_component_route(title: str, content: str, component_name: str, category: str = "general", component_data: Optional[Dict[str, Any]] = None, agent_id: str = "agent") -> SimpleEvent:
    """Create artifact component routing event for full-featured right panel components"""
    return SimpleEvent(
        type=SimpleEventType.ROUTE_TO_ARTIFACT,
        data={
            "title": title,
            "content": content,
            "category": category,
            "component_name": component_name,
            "component_data": component_data or {},
            "artifact_id": f"artifact_{int(time.time())}"
        },
        agent_name=agent_id
    )


def create_inline_component_route(content: str, component_name: str, component_data: Optional[Dict[str, Any]] = None, agent_id: str = "agent") -> SimpleEvent:
    """Create inline component routing event for lightweight UI elements in chat flow"""
    return SimpleEvent(
        type=SimpleEventType.ROUTE_TO_CHAT,
        data={
            "content": content,
            "component_name": component_name,
            "component_data": component_data or {}
        },
        agent_name=agent_id
    )


def create_ui_tool_action(tool_id: str, action_type: str, payload: Dict[str, Any]) -> SimpleEvent:
    """Create UI tool action event"""
    return SimpleEvent(
        type=SimpleEventType.UI_TOOL_ACTION,
        data={
            "tool_id": tool_id,
            "action_type": action_type,
            "payload": payload
        }
    )


def create_status_event(message: str) -> SimpleEvent:
    """Create status event"""
    return SimpleEvent(
        type=SimpleEventType.STATUS,
        data={"message": message}
    )


def create_error_event(error_message: str, error_code: Optional[str] = None) -> SimpleEvent:
    """Create error event"""
    data = {"message": error_message}
    if error_code:
        data["code"] = error_code
    
    return SimpleEvent(
        type=SimpleEventType.ERROR,
        data=data
    )
